### PR TITLE
Purge attachments on message delete

### PR DESCRIPTION
**Pull request by joshblum**
When a message explodes or is deleted with an attachment we would like to purge the attachment from local caches and attempt remote deletion.

### CODE REVIEW

File: go/chat/server.go [hunk 1]
```diff
@@ -1409,155 +1406,22 @@ func (h *Server) postAttachmentLocal(ctx context.Context, arg postAttachmentArg)
 		}
 	}
 
-	// get s3 upload params from server
-	params, err := h.remoteClient().GetS3Params(ctx, arg.ConversationID)
```
Comment (joshblum) [inline go/chat/server.go]: this method was unused so i removed cc @patrickxb

File: go/chat/server.go [hunk 2]
```diff
@@ -2101,30 +1965,6 @@ func (h *Server) attachmentMessage(ctx context.Context, conversationID chat1.Con
 
 }
 
-func (h *Server) deleteAssets(ctx context.Context, conversationID chat1.ConversationID, assets []chat1.Asset) {
```
Comment (joshblum) [inline go/chat/server.go]: also unused

File: go/chat/storage/storage.go [hunk 3]
```diff
@@ -442,6 +445,11 @@ func (s *Storage) updateAllSupersededBy(ctx context.Context, convID chat1.Conver
 		}
 	}
 
+	// conv source will queue asset deletions in the background
+	if convSource := s.G().ConvSource; convSource != nil {
```
Comment (mmaxim) [inline go/chat/storage/storage.go]: Why do this?

File: go/chat/attachment_httpsrv.go [hunk 4]
```diff
@@ -350,3 +386,34 @@ func (c *CachingAttachmentFetcher) FetchAttachment(ctx context.Context, w io.Wri
...
+	}
+
+	// get s3 params from server
+	s3params, err := ri().GetS3Params(ctx, convID)
+	if err != nil {
+		c.Debug(ctx, "error getting s3params: %s", err)
+		return err
+	}
+
+	// Delete the assets locally
```
[the remark this answers is not in the document]
Comment (joshblum) [reply go/chat/attachment_httpsrv.go]: ah good point, fixed

File: go/chat/storage/storage.go (hunk 3 above)
Comment (joshblum) [reply go/chat/storage/storage.go]: `ConvSource` isn't set on `g` in the storage tests. Should I add it to the test setup?

File: go/chat/attachment_httpsrv.go [hunk 5]
```diff
@@ -350,3 +386,34 @@ func (c *CachingAttachmentFetcher) FetchAttachment(ctx context.Context, w io.Wri
...
+	convID chat1.ConversationID, assets []chat1.Asset, ri func() chat1.RemoteInterface, signer s3.Signer) (err error) {
+	defer c.Trace(ctx, func() error { return err }, "DeleteAssets")()
+
+	if len(assets) == 0 {
+		return nil
+	}
+
+	// Delete the assets locally
+	for _, asset := range assets {
+		c.diskLRU.Remove(ctx, c.G(), c.cacheKey(asset))
```
Comment (joshblum) [inline go/chat/attachment_httpsrv.go]: os.Remove

File: go/chat/convsource.go [hunk 6]
```diff
@@ -42,6 +42,33 @@ func (s *baseConversationSource) SetRemoteInterface(ri func() chat1.RemoteInterf
 	s.ri = ri
 }
 
+// Sign implements github.com/keybase/go/chat/s3.Signer interface.
+func (s *baseConversationSource) Sign(payload []byte) ([]byte, error) {
```
Comment (joshblum) [inline go/chat/convsource.go]: make this a DeleteAsset interface to abstract from convsource

Comment (joshblum) [reply go/chat/convsource.go]: done

File: go/chat/attachment_httpsrv.go (hunk 5 above)
Comment (joshblum) [reply go/chat/attachment_httpsrv.go]: done

File: go/chat/attachment_httpsrv.go [hunk 7]
```diff
@@ -350,3 +386,35 @@ func (c *CachingAttachmentFetcher) FetchAttachment(ctx context.Context, w io.Wri
...
+	convID chat1.ConversationID, assets []chat1.Asset, ri func() chat1.RemoteInterface, signer s3.Signer) (err error) {
+	defer c.Trace(ctx, func() error { return err }, "DeleteAssets")()
+
+	if len(assets) == 0 {
+		return nil
+	}
+
+	// Delete the assets locally
+	for _, asset := range assets {
+		os.Remove(asset.Path)
```
Comment (mmaxim) [inline go/chat/attachment_httpsrv.go]: This is the remote path, I don't think this is going to work. Is there a test that checks to see if the file is removed?

Comment (joshblum) [reply go/chat/attachment_httpsrv.go]: yeah  :(  fixed & test added

File: go/chat/attachments/store.go [hunk 8]
```diff
@@ -317,19 +318,16 @@ func (a *Store) s3Conn(signer s3.Signer, region s3.Region, accessKey string) s3.
 }
 
 func (a *Store) DeleteAssets(ctx context.Context, params chat1.S3Params, signer s3.Signer, assets []chat1.Asset) error {
-	var errs []error
+
+	epick := libkb.FirstErrorPicker{}
```
Comment (mmaxim) [inline go/chat/attachments/store.go]: I need to use this guy more

**Pull request merged**